File: frontend/src/services/messageTracker.js
import { debug, warning } from './loggerService';

// Store for pending messages awaiting confirmation
const pendingMessages = {};

// Configuration
const CONFIRMATION_TIMEOUT = 500; // 0.5 seconds in milliseconds

// Interval ID for the checker
let checkerInterval = null;

/**
 * Track a message that was sent to the server
 * @param {object} messageData - The message that was sent
 */
export const trackMessage = (messageData) => {
  const id = messageData.id;
  if (!id) {
    debug('Cannot track message without ID');
    return;
  }
  
  pendingMessages[id] = {
    timestamp: Date.now(),
    data: messageData
  };
  
  // Start the checker if it's not already running
  ensureCheckerRunning();
};

/**
 * Confirm a message was received by the server
 * @param {string|number} messageId - The ID of the message to confirm
 * @returns {boolean} - Whether the message was found and confirmed
 */
export const confirmMessage = (messageId) => {
  if (messageId in pendingMessages) {
    const elapsed = (Date.now() - pendingMessages[messageId].timestamp) / 1000;
    debug(`Message ${messageId} confirmed after ${elapsed.toFixed(3)}s`);
    delete pendingMessages[messageId];
    return true;
  }
  return false;
};

/**
 * Check for expired messages and issue warnings
 */
const checkExpiredMessages = () => {
  const now = Date.now();
  const expiredIds = [];
  
  // Find expired messages
  Object.entries(pendingMessages).forEach(([id, info]) => {
    const elapsed = now - info.timestamp;
    if (elapsed > CONFIRMATION_TIMEOUT) {
      warning(`Message ${id} not confirmed after ${(elapsed/1000).toFixed(2)}s`);
      expiredIds.push(id);
    }
  });
  
  // Remove expired messages
  expiredIds.forEach(id => {
    delete pendingMessages[id];
  });
  
  // Stop checker if no more pending messages
  if (Object.keys(pendingMessages).length === 0 && checkerInterval) {
    clearInterval(checkerInterval);
    checkerInterval = null;
  }
};

/**
 * Ensure the message checker is running
 */
const ensureCheckerRunning = () => {
  if (!checkerInterval) {
    checkerInterval = setInterval(checkExpiredMessages, 100);
  }
};

// Export functions for testing
export const __pendingMessages = pendingMessages;

File: backend/src/utils/message_tracker.py
import threading
import time
from src.utils.logger import get_logger

logger = get_logger("message_tracker")
debug = logger.debug
warning = logger.warning
error = logger.error

# Dictionary to store messages waiting for confirmation
# Format: {message_id: {'timestamp': time_sent, 'data': message_data}}
pending_messages = {}

# Lock for thread-safe operations on the dictionary
message_lock = threading.Lock()

# Configuration
CONFIRMATION_TIMEOUT = 0.5  # Time in seconds before issuing warning

def track_message(message_data):
    id = message_data.get('id')
    with message_lock:
        pending_messages[id] = {
            'timestamp': time.time(),
            'data': message_data
        }

def confirm_message(message_id):
    with message_lock:
        if message_id in pending_messages:
            message_data = pending_messages.pop(message_id)
            elapsed = time.time() - message_data['timestamp']
            debug(f"Message {message_id} confirmed after {elapsed:.3f}s")
            return True
        return False

def check_expired_messages():
    """Check for and remove expired messages"""
    current_time = time.time()
    expired_messages = []
    
    # Identify expired messages
    with message_lock:
        for message_id, info in list(pending_messages.items()):
            elapsed = current_time - info['timestamp']
            if elapsed > CONFIRMATION_TIMEOUT:
                expired_messages.append((message_id, info))
    
    # Process expired messages (outside the lock to minimize lock time)
    for message_id, info in expired_messages:
        warning(f"Message {message_id} not confirmed after {CONFIRMATION_TIMEOUT:.2f}s")
        with message_lock:
            if message_id in pending_messages:  # Check again in case it was confirmed
                pending_messages.pop(message_id)

def start_tracker():
    """Start the background thread that checks for expired messages"""
    def tracker_thread():
        while True:
            try:
                time.sleep(0.1)  # Check frequently but not too often
                check_expired_messages()
            except Exception as e:
                error(f"Error in tracker thread: {e}")
    
    thread = threading.Thread(target=tracker_thread)
    thread.daemon = True
    thread.start()
    debug("Message tracker thread started")
    
# Start the tracker thread when the module is imported
start_tracker()